File: rust_exercises/leetcode/Problem750/Solution.java
class Solution {
    public int countCornerRectangles(int[][] grid) {
        int m = grid.length;
        int n = grid[0].length;
        
        int count = 0;
        
        Map<Integer, Integer>  map = new HashMap<>();
        for (int i = 0; i < m; i += 1) {
            for (int j1 = 0; j1 < n - 1; j1 += 1) {
                if (grid[i][j1] == 0) {
                    continue;
                }
                
                for (int j2 = j1 + 1; j2 < n; j2 += 1) {
                    if (grid[i][j2] == 0) {
                        continue;
                    }
                    int key = j1 * 200 + j2;
                    if (map.containsKey(key)) {
                        int prev = map.get(key);
                        count += prev;
                        map.replace(key, prev + 1);
                    } else {
                        map.put(key, 1);
                    }
                }
            }
        }
        return count;
    }
}

File: rust_exercises/leetcode/Problem750/solution.rs
use std::collections::HashMap;

impl Solution {
    pub fn count_corner_rectangles(grid: Vec<Vec<i32>>) -> i32 {
        let m = grid.len();
        let n = grid[0].len();
        
        let mut count = 0;
        let mut map: HashMap<(usize, usize), i32> = HashMap::new();
        
        for i in 0..m {
            for j1 in 0..(n - 1) {
                if grid[i][j1] == 0 {
                    continue;
                }
                
                for j2 in (j1 + 1)..n {
                    if grid[i][j2] == 0 {
                        continue;
                    }
                    
                    match map.get_mut(&(j1, j2)) {
                        Some(prev) => {
                            count += *prev;
                            *prev = *prev + 1;
                        },
                        None => { map.insert((j1, j2), 1); },
                    }
                }
            }
        }
        count
    }
}
